### ISSUE TITLE
[Bug] excessively deep TS error with multiple JSX children

### ISSUE THREAD

**Issue by CarsonF** [MadProbe/iterator-helpers-polyfill#14]
Anywhere there's multiple children given
```tsx
<div>
  <div />
  <div />
</div>
```
```
TS2589: Type instantiation is excessively deep and possibly infinite
```

I believe this is connected via fragments. They are iterable and children can be fragments.
```tsx
type ReactFragment = Iterable<ReactNode>;
```
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56623

Here's a reproduction if it helps.
https://codesandbox.io/s/react-typescript-template-forked-n1otpr?file=/src/index.tsx

I'm not sure what can be done to fix this, but it's preventing me from adopting the polyfill.
We are actually trying to use this in a node app, it just using jsx for rendering emails.

**Comment by MadProbe** [comment]
Thank you for reporting this bug.
The main problem is that `Iterator<T>.flatten<>()` had infinite recursion when it met `number` type, but I meant for it to show **every** possible combination of flattened iterable, ranging from no flattening to flattening every nested iterables.
I will soon publish a patch for this, but if you want the fix right now, do to `<project folder>/node_modules/iterator-helpers-polyfill/definitions/index.d.ts` and update the `FlattenIterator<T>` type to this:
```ts
type FlattenIterator<T, Depth extends number, Saved = never, IsFirst = true> = T extends string ? (number extends Depth ? Saved | T : never) : T extends _AsyncIteratorLike<infer V> ? number extends Depth ? 
    T extends Saved ? Saved : FlattenIterator<V, number, true extends IsFirst ? Saved : T | Saved, false> | (true extends IsFirst ? T : never)
    : 0 extends Decrement<Depth> ? V | Saved : FlattenIterator<V, Decrement<Depth>, Saved> : T;
```
P.S. Don't fret over this as it is something that requires quite sophisticated solution to provide good and elaborated typings for functions.

**Comment by CarsonF** [comment]
Thanks for the quick response. Surprised you've got a complex type fix ready to go unreleased haha.
This did fix the infinite loop bug in my codebase.

I'm curious, why you only union `T` after the first recurse?

**Comment by MadProbe** [comment]
> Thanks for the quick response. Surprised you've got a complex type fix ready to go unreleased haha. This did fix the infinite loop bug in my codebase.
> 
> I'm curious, why you only union `T` after the first recurse?

Maybe I could indeed write this type more efficient, but I prefer to not to break this monster of a code unless needed.
BTW, I have released version 2.3.0, which fixes this bug and some others

Edit: I have thought about it and I made it so it unions output type with `T` only when I have a `number` value type as the code in the case above only does all types down the `Iterable<...>` chain except the first type as it is pain in the ass to get it working otherwise and to not to break something in the process.
